### PR TITLE
Implement repair functionality for aliases colliding with indices bug

### PR DESCRIPTION
**Pull request by original-brownbear**
Follow up to #91456 implementing an automated fix for indices corrupted in 8.5.

Not sure this should get an automated test, could write one that corrupts the index in 8.5 and then does a rolling upgrade but maybe manually testing this is good enough in this case?

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 1]
```diff
@@ -2140,7 +2144,16 @@ public IndexMetadata build() {
             var aliasesMap = aliases.build();
             for (AliasMetadata alias : aliasesMap.values()) {
                 if (alias.alias().equals(index)) {
-                    throw new IllegalArgumentException("alias name [" + index + "] self-conflicts with index name");
+                    if (indexCreatedVersion.equals(Version.V_8_5_0)) {
+                        var updatedBuilder = ImmutableOpenMap.builder(aliasesMap);
+                        final var brokenAlias = updatedBuilder.remove(index);
+                        final var fixedAlias = AliasMetadata.newAliasMetadata(brokenAlias, index + "-alias-corrupted-by-8-5");
+                        aliasesMap = updatedBuilder.fPut(fixedAlias.getAlias(), fixedAlias).build();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2151]: Is there a risk that adding an alias with the same name as the index for an index created in 8.5 will now succeed and just result in the "corrupted" alias name instead?

Can we perhaps find or add an integration test that verifies that when adding aliases we have validation in place that catches this?

**Review by henningandersen — COMMENTED** [review]
> could write one that corrupts the index in 8.5 and then does a rolling upgrade

I think the level of testing here is adequate for ensuring it is fixed on upgrade.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2151]: Oh right ... this is quite annoying. I could obviously add the validation for this at another level as well (not sure where, it seems not entirely trivial to add this check elsewhere in a way that would only catch mutations) but it does add some extra complexity to the whole thing. Is it worth doing that for this corner case? The name collision checks are quite complicated as it is already and checking for this specifically would add yet more complexity at some other layer ...

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2151]: Could we add a `buildAndRepair` method that is used from `IndexMetadata.fromXContent`, passing down a boolean to a private `build` method in order to only do the repair from that path?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2151]: I was thinking about that, but the would mean that the fix will only apply with a full cluster restart and not with a rolling restart? Wouldn't that be painful for Cloud?

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2151]: We could also build and repair when receiving the index metadata over the wire maybe?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2151]: We could I guess :) lets do it.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:2156]: Added to from--x-content, reading from the wire and (not sure this is necessary but I found it weird for it to behave different form normal transport read) diff application now :)

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataTests.java [hunk 2]
```diff
@@ -504,11 +505,21 @@ public void testIndexAndAliasWithSameName() {
                 .numberOfShards(1)
                 .numberOfReplicas(0)
                 .putAlias(AliasMetadata.builder("index").build())
-                .build()
+                .build(randomBoolean())
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataTests.java:508]: Could we expand this test (or add one) to also verify that for indices created in v8.5, we reject an alias of the same name as the index?

**Review by henningandersen — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Henning!

**Pull request merged**